From b004e5f97c5ff4d1804029cef0d81aba87c07270 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?St=C3=A9phane=20Marchesin?= <marcheu@chromium.org>
Date: Wed, 15 Oct 2014 13:34:56 -0700
Subject: [PATCH 16/27] Fix workaround corner cases

This fixes two corner cases with the workaround:
- Fix the case where there is a depth but no stencil
- Fix the case there the depth mt hasn't been created

Change-Id: I9358076efeb227cd0ed62fbec71c04e193da8aac
---
 src/mesa/drivers/dri/i965/brw_misc_state.c | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/src/mesa/drivers/dri/i965/brw_misc_state.c b/src/mesa/drivers/dri/i965/brw_misc_state.c
index 4ee70bf..931e142 100644
--- a/src/mesa/drivers/dri/i965/brw_misc_state.c
+++ b/src/mesa/drivers/dri/i965/brw_misc_state.c
@@ -250,7 +250,7 @@ brw_workaround_depthstencil_alignment(struct brw_context *brw,
    brw->depthstencil.stencil_mt = NULL;
    if (depth_irb)
       brw->depthstencil.depth_mt = depth_mt;
-   if (stencil_irb)
+   if (stencil_irb && stencil_irb->mt)
       brw->depthstencil.stencil_mt = get_stencil_miptree(stencil_irb);
 
    /* Gen6+ doesn't require the workarounds, since we always program the
@@ -269,7 +269,7 @@ brw_workaround_depthstencil_alignment(struct brw_context *brw,
     * separate stencil.  To catch that case, we check whether
     * depth_mt->stencil_mt is non-NULL.
     */
-   if (depth_irb && invalidate_depth &&
+   if (depth_irb && depth_mt && invalidate_depth &&
        (_mesa_get_format_base_format(depth_mt->format) == GL_DEPTH_STENCIL ||
         depth_mt->stencil_mt)) {
       invalidate_depth = invalidate_stencil && depth_irb && stencil_irb
@@ -324,7 +324,7 @@ brw_workaround_depthstencil_alignment(struct brw_context *brw,
          tile_y = depth_irb->draw_y & tile_mask_y;
       }
 
-      if (stencil_irb) {
+      if (stencil_irb && stencil_irb->mt) {
          stencil_mt = get_stencil_miptree(stencil_irb);
          intel_miptree_get_image_offset(stencil_mt,
                                         stencil_irb->mt_level,
@@ -346,7 +346,7 @@ brw_workaround_depthstencil_alignment(struct brw_context *brw,
    }
 
    /* If we have (just) stencil, check it for ignored low bits as well */
-   if (stencil_irb) {
+   if (stencil_irb && stencil_irb->mt) {
       intel_miptree_get_image_offset(stencil_mt,
                                      stencil_irb->mt_level,
                                      stencil_irb->mt_layer,
@@ -449,7 +449,7 @@ brw_workaround_depthstencil_alignment(struct brw_context *brw,
                                              false);
       }
    }
-   if (stencil_irb) {
+   if (stencil_irb && stencil_irb->mt) {
       stencil_mt = get_stencil_miptree(stencil_irb);
 
       brw->depthstencil.stencil_mt = stencil_mt;
-- 
2.3.1

